python/Ganga/Utility/execute.py

```diff
@@ -184,24 +251,46 @@ def execute(command,
     if timed_out.isSet():
         return 'Command timed out!'
 
-    if update_env and 'env_output' in thread_output:
-        env.update(thread_output['env_output'])
+    # Decode any pickled objects from disk
+    if update_env:
+        update_env_thread.join()
+        if env_output_key in thread_output:
+            env.update(thread_output[env_output_key])
+        else:
+            logger.error("Expected to find the updated env after running a command")
+            logger.error("Command: %s" % command)
+            logger.error("stdout: %s" % stdout)
+            logger.error("stderr: %s" % stderr)
+            raise Exception("Missing update env after running command")
 
-    if not shell and 'pkl_output' in thread_output:
-        return thread_output['pkl_output']
+    if not shell:
+        update_pkl_thread.join()
+        if pkl_output_key in thread_output:
+            return thread_output[pkl_output_key]
+        else:
+            logger.error("Expected to find the pickled output after running a command")
+            logger.error("Command: %s" % command)
+            logger.error("stdout: %s" % stdout)
+            logger.error("stderr: %s" % stderr)
+            raise Exception("Missing pickled output after running command")
 
     try:
-        stdout = pickle.loads(stdout)
+        if isinstance(stdout, str) and stdout != '':
```

Comment:
is `stdout` ever not a `str`? Also, are we actually writing any pickled output to the `stdout` ever or does this always `except`?